### ISSUE TITLE
Pytest reevaluates walrus operations within assert statements

### ISSUE THREAD

**Issue by regananalytics** [pytest-dev/pytest#11115]
- [ ] a detailed description of the bug or problem you are having
- [ ] output of `pip list` from the virtual environment you are using
- [ ] pytest and operating system versions
- [ ] minimal example if possible

Ubuntu 22.04, pytest 7.3.2, python 3.10.11

Pytest 7.3.2 re-evaluates walrus operator statements if they appear in assert statements.

```
def test_walrus():
    assert (x := 2) == 2
    x  =  3
    assert x == 3, "This test fails because (x := 2) is reevaluated by pytest"
    
def test_walrus_no_assert():
    (x := 2) == 2
    x = 3
    assert x == 3, "This test passes because (x := 2) is not reevaluated"
```

**Comment by The-Compiler** [comment]
Bisected to 331bc1be4634cf95fdb33021d4355c35723b73fe:

- #11041

cc @aless10

**Comment by aless10** [comment]
Oh no! I'm thinking that a lot of edge cases are really really "edge", and I hope that nobody writes tests like this one. But I'm going to look at this. Is that ok with you?
Thanks

**Comment by regananalytics** [comment]
Yeah, this is obviously not a real test. I wrote it as a super simple and replicable example of the issue.

**Comment by The-Compiler** [comment]
@aless10 Sure, please go ahead! No worries, sometimes things can be unexpectedly tricky. It's much appreciated that you're taking care of those follow-up issues!
